### PR TITLE
fix: [SDK-5065] bound the crash-record cache on iOS

### PR DESCRIPTION
**Pull request by abdulraqeeb33**
# Description
## One Line Summary
Bounds the iOS crash-record cache (age ceiling, count and byte caps, write-size limit) using the retention policy shared from the KMP module.

Relates to [SDK-5065](https://linear.app/onesignal/issue/SDK-5065/remove-the-otel-observability-path-and-opentelemetry-dependency-from).

> [!WARNING]
> **Blocked on [OneSignal-KMP-SDK#23](https://github.com/OneSignal/OneSignal-KMP-SDK/pull/23)**, which is unmerged; the submodule is pinned to that branch head (`0513bf5`). When #23 merges the pin must be re-pointed at KMP `main` **and the suite re-run**: the shared policy changed behavior after this branch was verified, with eviction going from three tiers to four, protected names no longer claiming byte budget, and foreign-name dating tightened.

### Motivation

`FileLogStore` had no retention: no size limit on `save()`, no age ceiling on `listReadable()`, no count or byte cap, no eviction anywhere. A record that fails to upload is re-read and re-POSTed on every launch, indefinitely, while the directory grows unbounded. Android hit the same defect when OpenTelemetry's `disk-buffering` was removed, so rather than reimplement the policy here and let the two drift, it moved to `commonMain` as `CrashRetention`. This store now contributes only the file I/O.

### Scope

| Path | Behavior |
| --- | --- |
| `save()` | refuses payloads over the per-record limit, then trims to the caps, keeping the record just written |
| `listReadable()` | reclaims before materializing payloads, so an over-cap backlog is never loaded into memory |
| `deleteUnrecognizedEntries()` | also reclaims, since it is the only scan that runs when remote logging is disabled |

Non-obvious decisions:

**Unreadable attributes are the ordinary Apple case.** Data protection denies file attributes while the directory still lists, on every reboot until first unlock. The entry stays in the snapshot, since omitting it would put the file outside every bound at once, and a missing mtime is passed through as `nil` rather than a stand-in number: the policy reads whatever it is given back as an age, so the `0` an earlier revision substituted made every such record look maximally stale and deleted it. Age now falls back to the leading millis in the record's own name, and a record neither source can date is withheld from readers rather than uploaded or reaped, still bounded by the caps.

**The write suffix is derived from the policy, not a local constant.** Ownership is checked through that same `CrashRetentionPolicy`, so a local suffix that drifted would hide every brand-new record from every reader.

**In-flight names are reserved.** Concurrent crashing threads each hold a record open. The async reclaim skips in-flight names, and the write path unlinks whatever the selector returns without re-checking, so it reserves *every* in-flight name and not just the one it wrote.

**The temp sweep stays narrower than the shared selector.** `selectUnrecognized` reaps any non-owned file; iOS reaps only its own `.otlp.tmp`, having never run the OpenTelemetry pipeline, and files we did not write are not ours to assume about. It skips this store's in-flight temps and deletes per-entry, so one undeletable leftover cannot strand the pass.

The submodule pin moves `87e87fd` → `0513bf5`: the shared policy (KMP #20), bounded retry/backoff in the export path (#21), the logger-contract cleanup (#22), and nullable crash write times (#23).

# Testing
## Unit testing

`FileLogStoreRetentionTests` and `FileLogStoreRecordDatingTests` cover both caps and the age ceiling, the size limit at its exact boundary, in-flight records surviving a concurrent eviction and a racing temp sweep, and name-based dating when attributes are unreadable.

Full `OneSignalOSCore` suite green against the current pin: **135 tests, 0 failures** (`UnitTestApp_TestPlan_Reduced`, iPhone 17 Pro simulator). Perturbation-checked, both reverted: dropping the write-size guard reddens one case, and widening the store's bounds while the tests assert against `CrashRetention.shared.defaultPolicy` reddens five more. The policy decisions themselves are unit-tested in the shared module.

### CODE REVIEW

File: iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift [hunk 1]
```diff
@@ -133,38 +163,135 @@ final class FileLogStore: ILogFileStore {
...
+                message: "FileLogStore failed to reclaim \(name): \(error.localizedDescription)"
+            )
+        }
+    }
+
+    /// Snapshots the directory as the platform-neutral entries the shared policy consumes.
+    private func directoryEntries() throws -> [CrashDirEntry] {
+        try fileURLs().compactMap { url in
+            let values = try? url.resourceValues(forKeys: [.contentModificationDateKey, .fileSizeKey])
+            guard let modifiedAt = values?.contentModificationDate else {
```
[the remark this answers is not in the document]
Comment (abdulraqeeb33) [reply iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift]: Fixed in `cf50a0ac`. This was the right call, and the divergence from Android makes it sharper than stated: Android's `listEntries` uses `File.lastModified()`, which returns `0L` on failure rather than dropping the entry, so `nowMs - 0 > maxReadAgeMillis` is trivially true and Android reclaims the file. Same shared policy, opposite outcomes — Android cleans up, iOS leaked for the life of the install.

`directoryEntries()` now dates an unreadable attribute to the epoch instead of omitting it, so iOS reaches the same outcome. `compactMap` became `map`, since nothing returns nil any more and the optional-flattening was misleading about whether entries could still vanish.

On the missing-size half: no change needed. Android builds `lengthBytes` from `File.length()`, which also yields `0` when the length cannot be determined, so both platforms already agree that an unsized record claims no budget. The asymmetry was only ever in the date — Java folds that failure into a sentinel while `URLResourceValues` folds it into `nil`, and the old code escalated `nil` into dropping the whole record.

Covered by `testRecordWithUnreadableAttributesIsReclaimedRatherThanLeaked`, verified red against the old `return nil`.

File: iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift [hunk 2]
```diff
@@ -133,38 +163,135 @@ final class FileLogStore: ILogFileStore {
...
+        }
+        let overflow = CrashRetention.shared.selectOverflowOwned(
+            entries: entries,
+            nowMs: Self.nowMillis(),
+            keepName: keepName,
+            policy: Self.retentionPolicy
+        )
+        for entry in overflow {
+            remove(name: entry.name)
+        }
```
[the remark this answers is not in the document]
Comment (abdulraqeeb33) [reply iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift:295]: Deliberate, and now commented at `enforceAccumulationCaps` — thanks for pushing on it. The split matches pre-deletion Android: the write path runs overflow only so the crashing thread does not pay for an expiry scan when nothing is over cap, and under-cap expired records are reclaimed on the next uploader pass by `listReadable` / `deleteUnrecognizedEntries`.

Behavior unchanged, but this is the second independent review to question it against the `ILogFileStore` wording, which is fair evidence the intent was not legible in the code. The comment now says why and where expiry actually happens.

File: iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift [hunk 3]
```diff
@@ -56,11 +67,23 @@ final class FileLogStore: ILogFileStore {
...
+                    + "\(Self.retentionPolicy.maxRecordBytes)-byte limit"
+            )
+            return false
+        }
         do {
             try createRootDirectory()
             let timestamp = Int64(Date().timeIntervalSince1970 * 1_000)
             let id = "\(timestamp)-\(UUID().uuidString)\(Self.ownedFileSuffix)"
             try writeDurably(bytes.data, to: rootURL.appendingPathComponent(id))
+            enforceAccumulationCaps(keepName: id)
```
[the remark this answers is not in the document]
Comment (abdulraqeeb33) [reply iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift]: Accepted trade-off, consistent on both platforms. `save()` must finish synchronously on the crashing thread, and deferring the trim to `ioQueue` would mean a crash that kills the process before the queue drains leaves the directory over cap — the failure mode the caps exist to prevent.

The cheap-exit you suggest is already there: `isWithinCaps` short-circuits before any sort or unlink, so the steady-state cost is one listing rather than a listing plus sort. The unbounded case is an inherited pre-retention backlog, which is bounded once by the first post-upgrade crash and cheap from then on. Android made the same call, and keeping the two in step matters more here than shaving the first-crash cost.

**Comment by fadi-george** [comment]
I found a few issues worth addressing before merge:

1. Unreadable timestamps become `0`, making an in-flight `.otlp.tmp` appear old enough for immediate deletion. Unknown age should preserve temporary files while remaining stale for owned-record retention.

2. `fileURLs()` still drops entries when `isRegularFile` cannot be read, before the new attribute fallback runs. Keep entries unless `isRegularFile == false`, and cover this path through the injected lookup.

3. New `save()` failure branches call `OneSignalLog` during uncaught-exception handling. That can invoke application listeners or remote logging. Please use a crash-safe sink or suppress these diagnostics.

4. Synchronous save eviction can race with asynchronous reclaim. At minimum, treat an already-removed file as a successful cleanup; ideally protect in-flight record names.

**Comment by abdulraqeeb33** [comment]
Addressed in 1e9949f4:

1. Unknown mtime stays epoch for owned records (still stale / reclaimed) but the temp sweep now skips `lastModifiedMs == 0`, so an in-flight `.otlp.tmp` is not treated as ancient.
2. Entries are kept unless `isRegularFile == false`. That bit goes through the injected lookup so the nil path is testable.
3. `save()` refusal and crash-thread eviction log through `OSCrashLogger` (`NSLog`), not `OneSignalLog`.
4. `NSFileNoSuchFileError` / ENOENT counts as a successful cleanup. The name being written (and its `.tmp`) is held out of the async sweeper for the duration of `save()`.

Covered by new cases in `FileLogStoreRetentionTests`; verified red against the old code.

**Comment by fadi-george** [comment]
`1e9949f` addresses the async reclaim/temp-sweep races, but the synchronous write-path cap enforcement still ignores other in-flight names. It also intentionally treats unknown-mtime owned records as expired, so the potential post-reboot data loss remains.

**Comment by abdulraqeeb33** [comment]
Both points were right, and the second one is worse than you framed it. Fix is up as [OneSignal-KMP-SDK#23](https://github.com/OneSignal/OneSignal-KMP-SDK/pull/23). **This PR should not merge until that lands and is adopted here** — as it stands it ships the data loss.

## The unknown-mtime case destroys live crash reports

Not just "potential." With `lastModifiedMs` falling back to `0`, `selectExpiredOwned` evaluates `nowMs - 0 > maxReadAgeMillis`, which is trivially true, so the record is deleted. The test I added asserts exactly that, on a record **60 seconds old**:

```swift
try writeRecord(named: "opaque.otlp", ageMillis: 60_000)
...
XCTAssertFalse(fileExists("opaque.otlp"))
```

So a crash captured before a reboot is destroyed on the first post-reboot launch instead of uploaded, and my test pinned that as correct. My reasoning error was justifying it as Android parity: on Android `File.lastModified() == 0` means stat failed, whereas on iOS unreadable attributes are a routine transient state with a known end at first unlock. Matching Android's disposal for a condition that resolves itself is what turned a leak fix into data loss.

Worth keeping the other constraint in view, since it's why the naive fix is wrong: the previous behavior *omitted* these entries from the snapshot, which put them outside every bound at once — uncounted by the caps, unselectable by either reclaim pass, leaked for the life of the install. Any fix has to be bounded **and** non-destructive.

## The fix dates records by their own filename

Rather than an "unknown" state, #23 recovers the write time from the record's name. Records are written as `{millis}-{uuid}.otlp` from the same clock reading that stamps the file, and `CrashRetention` already had `leadingMillis` — it was just used as an overflow tie-break and never consulted for age. Age now resolves through one helper:

```kotlin
fun effectiveWriteTimeMs(entry: CrashDirEntry): Long? =
    entry.lastModifiedMs?.takeIf { it > 0 } ?: leadingMillis(entry.name)
```

The filesystem wins when readable, so the normal path decides exactly as before. This beats an unknown-only design in both directions: a 60-second-old record with unreadable attributes survives, **and** a genuinely 80-hour-old one still expires. Under unknown-only the latter would have survived until overflow pressure, quietly suspending the age ceiling — trading the data loss for a retention leak.

`CrashDirEntry.lastModifiedMs` became `Long?` deliberately, despite `KotlinLong?` being clumsier at Swift call sites. It means `now - entry.lastModifiedMs` no longer compiles, so every remaining age gate on both platforms *must* be revisited. A sentinel would have kept them compiling and silently wrong, which is the same invisibility that produced this bug.

Verified by perturbation: restoring `?: 0L` turns 9 tests red, including `selectExpiredOwned_retains_a_recent_record_whose_timestamp_is_unreadable`.

## In-flight names

Also fixed in #23. `selectOverflowOwned` now takes `keepNames: Set<String>` instead of a single name, so concurrent saves each protect their own in-flight record. Documented rule: protected names are never evicted even when they alone exceed the caps — deleting a file another thread has open corrupts the crash being captured right now, the set is bounded by concurrently crashing threads, and the overshoot ends when those writes finish. Budget claims stay capped per record, so an oversized protected record can't start the budget over cap and wipe the backlog.

## Android

I had the Android path checked too, since it feeds the same policy. The deletion chain is identical and reachable in code, but not realistically triggerable in production: `File.lastModified()` returns `0L` only on I/O failure, not as an attributes-unavailable sentinel, the crash directory is in credential-encrypted `cacheDir`, and `OneSignalImp` bails out when `isAndroidUserUnlocked` is false, so no reclaim pass runs before first unlock. Nothing there is Direct Boot aware. Edge-case I/O error at worst, not a systematic class like iOS. It's covered by the same fix regardless, and the non-positive guard protects Android even before it adopts.

Android also has no test for zero or unknown mtime at all — its fixture helper always sets a real timestamp — so that gap is being closed as part of the adoption.

## What lands here after #23

`directoryEntries()` drops the `?? 0`, the age filters move onto the shared helper, `reclaim()` and `enforceAccumulationCaps` pass `keepNames`, and the doc comment above `directoryEntries()` gets rewritten — it currently states the opposite of the new contract. The test quoted above gets inverted to assert survival, and its fixtures move to production-shaped `{millis}-{uuid}.otlp` names so the fallback is actually exercised rather than sidestepped.

One incidental improvement: the temp sweep loses its hand-rolled `lastModifiedMs > 0` sentinel in favour of the shared helper, and since `.otlp.tmp` names parse, interrupted writes become reapable before first unlock instead of being skipped.

**Comment by abdulraqeeb33** [comment]
Both of your points are now implemented on this branch, not just planned. `cfecade9` (pin) and `38823255` (code and tests). All four checks green.

**Unknown-mtime data loss.** Records are now dated by their own `{millis}-{uuid}.otlp` name when the filesystem timestamp is unreadable, so a minute-old crash captured before a reboot survives the pre-first-unlock launch and uploads. `testRecordWithUnreadableAttributesIsDatedByItsNameAndUploaded` replaces the test that asserted the deletion as correct.

One detail worth recording, because it changes where the fix actually lives: removing the `?? 0` from `directoryEntries()` is **inert on its own**. KMP's `takeIf { it > 0 }` already normalizes a fabricated zero, so the whole suite stays green with the `?? 0` restored. Perturbing to `?? 1` reproduces the original deletion exactly — so any *positive* invented number restores the bug, and the load-bearing change is in KMP. The iOS change is the store no longer asserting something it does not know.

**Write-path in-flight names.** Fixed here rather than deferred. `enforceAccumulationCaps` now passes `snapshotInFlightNames().union([keepName])`, reusing the `NSLock`-guarded registry that already existed for the temp sweep.

This one turned out to be real deletion, not cap drift. `reclaim` filters its loop with `where !inFlight.contains(...)`, but the write path calls `remove(name:crashSafe: true)` on whatever the selector returns with no in-flight check of its own, so a sibling thread's just-renamed record was genuinely unlinked. Pinned by `testSaveDoesNotEvictAnotherThreadsInFlightRecord`, confirmed red against single-name protection.

Android does **not** have this yet — no equivalent registry — so the two platforms currently diverge on the shared contract. Tracked as SDK-5129, scoped to Android.

**Still gated on [KMP #23](https://github.com/OneSignal/OneSignal-KMP-SDK/pull/23), and the re-point is not mechanical.** This branch pins `0513bf5`; #23 has since moved to `215ff06` with three behavior changes — protected names no longer claim shared byte budget (previously four in-flight records at the ceiling returned the *entire* backlog for eviction), the non-positive screen now covers name-derived times with foreign names datable only when entirely numeric, and eviction went from three tiers to four so an expired record no longer outranks an undatable one.

So when #23 merges this needs a re-pin **and** a full re-run, not just a pin bump. The eviction-ordering tests here are the most likely to be affected by the tier change.

**Comment by fadi-george** [comment]
Id merge kmp first and update submodule to kmp main

**Review by fadi-george — APPROVED** [review]

File: iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift [hunk 4]
```diff
@@ -353,20 +314,10 @@ final class FileLogStore: ILogFileStore {
...
-    /// whatever it is given back as an age, so the `0` this used to pass made every record
-    /// with unreadable attributes look maximally stale and had it deleted — on Apple platforms
-    /// that is every record in the directory, on every reboot, until first unlock. Age is
-    /// resolved by `CrashRetention.effectiveWriteTimeMs`, which recovers the write time from
-    /// the leading millis in the record's own name; only an entry neither source can date is
-    /// undatable, and that is bounded by the caps rather than by expiry.
-    ///
-    /// Entries are dropped only when `isRegularFile` is known false, not when that bit cannot
-    /// be read — the latter used to discard the file before this fallback could run.
+    /// Attributes are routinely unreadable under data protection before first unlock, so the
```
Comment (nan-li) [inline iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift:317]: non-blocking but more of a train of thought: the last commits making comments shorter is helpful but shorter isn't necessarily more understandable or clearer to a human reader. 

For tests, I would argue the cut comment lines should actually stay, because a test comment is sometimes the only place we know why something exists, or where it regressed before and how.

File: iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift [hunk 5]
```diff
@@ -294,19 +264,12 @@ final class FileLogStore: ILogFileStore {
         return withheld
     }
 
-    /// Trims the directory back inside the accumulation caps after a write, always keeping
-    /// [keepName]. Runs synchronously on the crashing thread, so it exits on one directory
-    /// listing in the steady state and only sorts when the caps are actually breached.
+    /// Trims back inside the accumulation caps after a write, reserving every in-flight name and
```
Comment (nan-li) [inline iOS_SDK/OneSignalSDK/OneSignalOSCore/Source/Logging/FileLogStore.swift:267]: more trains of thought as a human reviewer: this method and the comment was hard to parse. Tracing the call and asking AI for explanation is how I was able to understand what this method is for. I think a key idea is this is running in a crashing thread and needs to be as light as possible: with the overflow check only. For a human that is revisiting this method, `trimAfterCrashWrite` or `trimOnCrashPath` might be a clearer name

**Review by nan-li — COMMENTED** [review]
I don't really have any comments that the review bots haven't already been helpful on. Maybe more like a high-level trains of thought about readability to human reviewers or a future human reader.

Also any intuition on how big is a real iOS crash record is? We've taken the android numbers but if it's a few kb, then 50 records is well under 2 MB and that cap would never trigger. If it's like 400 KB, we would keep four crashes and the count cap wouldn't take any effect.
